File: todo/src/main/java/com/johnny/rxfluxtodo/TodoRecyclerAdapter.java
package com.johnny.rxfluxtodo;
/*
 * Copyright (C) 2017 Johnny Shieh Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import com.johnny.rxfluxtodo.action.TodoActionCreator;
import com.johnny.rxfluxtodo.model.Todo;

import android.support.v7.widget.RecyclerView;
import android.text.SpannableString;
import android.text.style.StrikethroughSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.TextView;

import java.util.List;


/**
 * description
 *
 * @author Johnny Shieh (JohnnyShieh17@gmail.com)
 * @version 1.0
 *
 * Created on 2017/3/28
 */
public class TodoRecyclerAdapter extends RecyclerView.Adapter<TodoRecyclerAdapter.ViewHolder> {

    private TodoActionCreator mActionCreator;
    private List<Todo> mTodoList;

    public TodoRecyclerAdapter(TodoActionCreator actionsCreator) {
        mActionCreator = actionsCreator;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int i) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.todo_row_layout, parent, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolder, int i) {
        viewHolder.bindView(mActionCreator, mTodoList.get(i));
    }

    @Override
    public int getItemCount() {
        return null == mTodoList ? 0 : mTodoList.size();
    }

    public void setItems(List<Todo> todoList) {
        mTodoList = todoList;
        notifyDataSetChanged();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        public TextView todoText;
        public CheckBox todoCheck;
        public Button todoDelete;

        public ViewHolder(View v) {
            super(v);
            todoText = (TextView) v.findViewById(R.id.row_text);
            todoCheck = (CheckBox) v.findViewById(R.id.row_checkbox);
            todoDelete = (Button) v.findViewById(R.id.row_delete);

        }

        public void bindView(final TodoActionCreator actionCreator, final Todo todo) {
            if (todo.isComplete()) {
                SpannableString spanString = new SpannableString(todo.getText());
                spanString.setSpan(new StrikethroughSpan(), 0, spanString.length(), 0);
                todoText.setText(spanString);
            } else {
                todoText.setText(todo.getText());
            }


            todoCheck.setChecked(todo.isComplete());
            todoCheck.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    actionCreator.toggleComplete(todo);
                }
            });

            todoDelete.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    actionCreator.destroy(todo.getId());
                }
            });
        }
    }
}

File: todo/src/main/java/com/johnny/rxfluxtodo/TodoRecyclerAdapter.kt
package com.johnny.rxfluxtodo

/*
 * Copyright (C) 2017 Johnny Shieh Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import android.text.SpannableString
import android.text.style.StrikethroughSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.johnny.rxfluxtodo.model.Todo
import com.johnny.rxfluxtodo.todo.TodoActionCreator
import kotlinx.android.synthetic.main.todo_row_layout.view.*


/**
 * description
 *
 * @author Johnny Shieh (JohnnyShieh17@gmail.com)
 * @version 1.0
 *
 * Created on 2017/3/28
 */
class TodoRecyclerAdapter(private val mActionCreator: TodoActionCreator) :
    RecyclerView.Adapter<TodoRecyclerAdapter.ViewHolder>() {

    private var mTodoList: List<Todo>? = null

    override fun onCreateViewHolder(parent: ViewGroup, i: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.todo_row_layout, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, i: Int) {
        viewHolder.bindView(mActionCreator, mTodoList!![i])
    }

    override fun getItemCount(): Int {
        return if (null == mTodoList) 0 else mTodoList!!.size
    }

    fun setItems(todoList: List<Todo>) {
        mTodoList = todoList
        notifyDataSetChanged()
    }

    class ViewHolder(v: View) : RecyclerView.ViewHolder(v) {

        fun bindView(actionCreator: TodoActionCreator, todo: Todo) {
            if (todo.isComplete) {
                val spanString = SpannableString(todo.text)
                spanString.setSpan(StrikethroughSpan(), 0, spanString.length, 0)
                itemView.tvTodoContent.text = spanString
            } else {
                itemView.tvTodoContent.text = todo.text
            }

            itemView.chkComplete.isChecked = todo.isComplete
            itemView.chkComplete.setOnClickListener { actionCreator.toggleComplete(todo) }

            itemView.btnDelete.setOnClickListener { actionCreator.destroy(todo.id) }
        }
    }
}
